src/main/java/org/elasticsearch/index/store/FsDirectoryService.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Collections;
+import java.util.Set;
+
+import com.google.common.collect.Sets;
+import org.apache.lucene.store.*;
+import org.apache.lucene.util.Constants;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardPath;
+import org.elasticsearch.index.store.DirectoryService;
+import org.elasticsearch.index.store.IndexStore;
+import org.elasticsearch.index.store.IndexStoreModule;
+import org.elasticsearch.index.store.StoreException;
+
+/**
+ */
+public class FsDirectoryService extends DirectoryService implements StoreRateLimiting.Listener, StoreRateLimiting.Provider {
+
+    protected final IndexStore indexStore;
+
+    private final CounterMetric rateLimitingTimeInNanos = new CounterMetric();
+    private final ShardPath path;
+
+    @Inject
+    public FsDirectoryService(@IndexSettings Settings indexSettings, IndexStore indexStore, ShardPath path) {
+        super(path.getShardId(), indexSettings);
+        this.path = path;
+        this.indexStore = indexStore;
+    }
+
+    @Override
+    public long throttleTimeInNanos() {
+        return rateLimitingTimeInNanos.count();
+    }
+
+    @Override
+    public StoreRateLimiting rateLimiting() {
+        return indexStore.rateLimiting();
+    }
+
+    protected final LockFactory buildLockFactory() throws IOException {
+        String fsLock = indexSettings.get("index.store.fs.lock", indexSettings.get("index.store.fs.fs_lock", "native"));
+        LockFactory lockFactory;
+        if (fsLock.equals("native")) {
+            lockFactory = NativeFSLockFactory.INSTANCE;
+        } else if (fsLock.equals("simple")) {
+            lockFactory = SimpleFSLockFactory.INSTANCE;
+        } else {
+            throw new StoreException(shardId, "unrecognized fs_lock \"" + fsLock + "\": must be native or simple");
+        }
+        return lockFactory;
+    }
+
+    @Override
+    public Directory newDirectory() throws IOException {
+        final Path location = path.resolveIndex();
+        Files.createDirectories(location);
+        Directory wrapped = newFSDirectory(location, buildLockFactory());
+        return new RateLimitedFSDirectory(wrapped, this, this) ;
+    }
+
+
+    @Override
+    public void onPause(long nanos) {
+        rateLimitingTimeInNanos.inc(nanos);
+    }
+
+    /*
+    * We are mmapping docvalues as well as term dictionaries, all other files are served through NIOFS
+    * this provides good random access performance while not creating unnecessary mmaps for files like stored
+    * fields etc.
+    */
+    private static final Set<String> PRIMARY_EXTENSIONS = Collections.unmodifiableSet(Sets.newHashSet("dvd", "tim"));
+
+
+    protected Directory newFSDirectory(Path location, LockFactory lockFactory) throws IOException {
```

Comment:
isn't it still a problem? since on Lucene, it will create Mmap for windows, but we don't want to file switch to NIO on windows?